Move preliminary uvflag feature from hera_qm to pyuvdata

adampbeardsley opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Moved UVFlag class from hera_qm to pyuvdata. UVFlag is currently _not_ a UVBase object, but we intend to eventually upgrade to such. We would like to move it to pyuvdata "as-is" as a non-documented feature to help development of other codes that use it but do not want to require hera_qm.
We also moved relevant utility functions, and all the associated tests (for utils and UVFlag).

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Provides an object for handling flags, removes hera_qm dependency from other code bases.
<!--- If it fixes an open issue, please link to the issue here. If this PR closes an issue, put the word 'closes' before the issue link to auto-close the issue when the PR is merged. -->
Addresses first point of issue #548.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] My change introduces new functionality that should be highlighted in the tutorial.
  - [ ] I have updated the tutorial accordingly.
- [ ] If this is a bug fix, it includes a new test that breaks as a result of the bug (if possible)
- [ ] If this is a breaking change, it includes backwards compatibility and deprecation warnings (if possible)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests pass in both python 2 and python 3.
- [x] My change requires an update to the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md).
  - [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/master/CHANGELOG.md) accordingly.

bhazelton commented on pyuvdata/tests/__init__.py:
```diff
@@ -201,6 +201,13 @@ def skipIf_no_h5py(test_func):
 
 
+def skipIf_no_hera_cal(test_func):
+    """defines a decorator to skip tests that require h5py."""
```
`h5py` -> `hera_cal`

bhazelton commented on pyuvdata/utils.py:
```diff
@@ -959,3 +959,123 @@ def _reraise_context(fmt, *args):
+        axis (int, tuple, optional): Axis or axes to collapse. Default is all.
+        returned (Bool): Whether to return sum of weights. Default is False.
+    '''
+    if alg == 'mean':
```
Might be cleaner to define a dict with keys equal to the algorithm names and values equal to the functions. Then you can use `getattr` to call the appropriate function. For a similar example, see the `interp` method on UVBeam.

bhazelton commented on pyuvdata/utils.py:
```diff
@@ -959,3 +959,123 @@ def _reraise_context(fmt, *args):
+def collapse(a, alg, weights=None, axis=None, returned=False):
+    ''' Parent function to collapse an array with a given algorithm.
+    Args:
+        a (array): Input array to process.
```
I'm slightly opposed to single letter variable names. Why not just use `array`?

bhazelton commented on pyuvdata/utils.py:
```diff
@@ -959,3 +959,123 @@ def _reraise_context(fmt, *args):
+        weights (array, optional): weights for collapse operation (e.g. weighted mean).
+            NOTE: Some subfunctions do not use the weights. See corresponding doc strings.
+        axis (int, tuple, optional): Axis or axes to collapse. Default is all.
+        returned (Bool): Whether to return sum of weights. Default is False.
```
this is not a super intuitive keyword name for me. Maybe `return_weights`?

bhazelton commented on pyuvdata/uvflag.py:
```diff
@@ -0,0 +1,832 @@
+
+    def to_flag(self):
+        '''Convert to flag mode. NOT SMART. Simply removes metric_array and initializes
+        flag_array with Falses.
```
Is there a method to threshold a metric to turn it into a flag array?

adampbeardsley replied:
Currently, this is done in `xrfi`, and the function here is just for converting the UVFlag mode. We could implement a very simple version here (a `threshold` keyword?), but I don't think we want to get into the business of flagging algorithms. What do you think?

bhazelton replied:
yeah, this probably belongs in analysis code. If someone wants it in the future it can be added then.

bhazelton replied:
A very simple "threshold" version would also be totally reasonable.

bhazelton commented on pyuvdata/tests/test_uvflag.py:
```diff
@@ -0,0 +1,1280 @@
+    hd = HERAData(test_d_file, filetype='miriad')
+    hd.read()
+    uvf2 = UVFlag(hd)
+    nt.assert_equal(uvf1, uvf2)
```
What utility are you testing here that requires hera_cal? It seems like all of this can be done with just pyuvdata...

adampbeardsley replied:
This will be moved back to hera_qm.

mwilensky768 commented:
Another note is that, as is, install requires h5py.

bhazelton commented:
I think it's ok to add this "as-is" as a non-documented feature, but we should be sure to make all the relevant issues to track the needed improvements and documentation in conjunction with merging this PR.

adampbeardsley commented on pyuvdata/uvflag.py:
```diff
@@ -0,0 +1,832 @@
+            raise ValueError('Must pass in UVData object or UVFlag object of type '
+                             '"baseline" to match.')
+        if self.type != 'waterfall':
+            raise ValueError('Cannot convert from type "' + self.type + '" to "baseline".')
```
We made an issue (#565 ) to address this after we merge this PR.

adampbeardsley commented on pyuvdata/uvflag.py:
```diff
@@ -0,0 +1,832 @@
+                del(fobj)
+        elif isinstance(input, UVFlag):
+            self.type = input.type
+            self.history = input.history + 'Copied by ' + self.pyuvdata_version_str + history
```
If you're ok with it, we'd like to punt on the version string (made issue #568). We got around the second comment by simply removing the option to instantiate a UVFlag with another UVFlag (functionality remains with uvflag.copy)

bhazelton replied:
yep, sounds good. Thanks for making the issue.

adampbeardsley commented:
Ok @bhazelton I think we've addressed your comments (either changes to the code, or punting with associated issues). Tests are passing, and we updated the CHANGELOG, so I think we're ready for a real review.

bhazelton commented on pyuvdata/uvflag.py:
```diff
@@ -0,0 +1,832 @@
+from . import version as uvversion
+from . import telescopes as uvtel
+import warnings
+import h5py
```
I think we agreed to move these imports inside the function where they're needed and to decorate tests that require h5py with the `skipIf_no_healpy` decorator.
